### PR TITLE
feat(lsp): execute command "ast-grep.applyAllFixes"

### PR DESCRIPTION
**Pull request by SoonIter**
https://github.com/ast-grep/ast-grep-vscode/pull/261

## Summary by CodeRabbit

- **New Features**
	- Added a new custom command for applying all fixes in the Language Server Protocol (LSP) service.
	- Introduced testing capabilities for LSP functionality including initialization, code actions requests, and command execution.

- **Improvements**
	- Enhanced diagnostic handling to support conditional returns.
	- Optimized fix application process by refining the parameter requirements.

- **Dependencies**
	- Added `serde_json` to improve JSON handling capabilities.

### CODE REVIEW

File: crates/lsp/src/lib.rs [hunk 1]
```diff
@@ -198,6 +205,14 @@ impl<L: LSPLang> LanguageServer for Backend<L> {
       .await;
     Ok(self.on_code_action(params).await)
   }
+
+  async fn execute_command(&self, params: ExecuteCommandParams) -> Result<Option<Value>> {
+    self
+      .client
+      .log_message(MessageType::INFO, "execute command!")
```
Comment (HerringtonDarkholme) [inline crates/lsp/src/lib.rs:212]: can we have a more descriptive message?

File: crates/lsp/src/lib.rs [hunk 2]
```diff
@@ -282,30 +310,39 @@ impl<L: LSPLang> Backend<L> {
...
     let mut diagnostics = vec![];
     let path = uri.to_file_path().ok()?;
 
-    let rules = match &self.rules {
-      Ok(rules) => rules.for_path(&path),
-      Err(_) => {
-        return Some(());
-      }
-    };
+    let rules = self.rules.as_ref().ok()?.for_path(&path);
```
Comment (HerringtonDarkholme) [inline crates/lsp/src/lib.rs:321]: This is not the same as the previous code. 
if rules is Err then the get_diagnosticsw returns None

File: crates/lsp/src/lib.rs [hunk 3]
```diff
@@ -446,158 +490,61 @@ impl<L: LSPLang> Backend<L> {
...
+    match command.as_ref() {
+      APPLY_ALL_FIXES => {
+        self
+          .client
+          .log_message(
+            MessageType::LOG,
+            format!("Running ExecuteCommand {}", command),
+          )
+          .await;
+        let uri = Url::parse(arguments.first()?["uri"].as_str()?).ok()?;
```
Comment (HerringtonDarkholme) [inline crates/lsp/src/lib.rs:515]: This code looks pretty weird. We should add a serde struct to decode the JSON to a Rust struct?

File: crates/lsp/src/lib.rs (hunk 1 above)
Comment (SoonIter) [reply crates/lsp/src/lib.rs:212]: it is copied from the code of code_action
```rust
  async fn code_action(&self, params: CodeActionParams) -> Result<Option<CodeActionResponse>> {
    self
      .client
      .log_message(MessageType::INFO, "run code action!")
      .await;
    Ok(self.on_code_action(params).await)
  }
```
more descriptive message is shown here

https://github.com/SoonIter/ast-grep/blob/9206a9c154e7d7bd34345424753257bb649a3517/crates/lsp/src/lib.rs#L463-L469

File: crates/lsp/src/lib.rs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 use dashmap::DashMap;
 use serde::{Deserialize, Serialize};
+use serde_json::Value;
```
[the remark this answers is not in the document]
Comment (SoonIter) [reply crates/lsp/src/lib.rs:3]: ？？？

File: crates/lsp/src/lib.rs [hunk 5]
```diff
@@ -192,12 +199,12 @@ impl<L: LSPLang> LanguageServer for Backend<L> {
...
-    self
-      .client
-      .log_message(MessageType::INFO, "run code action!")
-      .await;
     Ok(self.on_code_action(params).await)
   }
+
+  async fn execute_command(&self, params: ExecuteCommandParams) -> Result<Option<Value>> {
+    Ok(self.on_execute_command(params).await)
+  }
```
[the remark this answers is not in the document]
Comment (SoonIter) [reply crates/lsp/src/lib.rs:207]: In general, params will not be wrong

File: crates/lsp/src/lib.rs [hunk 6]
```diff
@@ -282,30 +302,44 @@ impl<L: LSPLang> Backend<L> {
...
       diagnostics.extend(ms.into_iter().map(to_diagnostic));
     }
+    Some(diagnostics)
+  }
+
+  async fn publish_diagnostics(&self, uri: Url, versioned: &VersionedAst<StrDoc<L>>) -> Option<()> {
+    let diagnostics = self
+      .get_diagnostics(&uri, versioned)
+      .await
+      .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (SoonIter) [reply crates/lsp/src/lib.rs:335]: too much noise log

File: crates/lsp/src/lib.rs [hunk 7]
```diff
@@ -446,158 +487,84 @@ impl<L: LSPLang> Backend<L> {
...
-              "character": 16
-            }
-          },
-          "code": "no-console-rule",
-          "source": "ast-grep",
-          "message": "No console.log"
-        }
-      ],
-      "only": ["source.fixAll"]
+      _ => None,
```
[the remark this answers is not in the document]
Comment (SoonIter) [reply crates/lsp/src/lib.rs:567]: too old, code already changed

File: crates/lsp/src/lib.rs [hunk 8]
```diff
@@ -446,158 +487,64 @@ impl<L: LSPLang> Backend<L> {
...
```
[the remark this answers is not in the document]
Comment (SoonIter) [reply crates/lsp/src/lib.rs]: looks good

**Pull request merged**